horovod/common/operations.cc

```diff
@@ -854,7 +870,63 @@ void CoordinateCacheAndState(CacheCoordinator& cache_coordinator,
 //      otherwise we may end up dispatching many blocked threads and never make
 //      progress if we have a thread pool limit.
 bool RunLoopOnce(HorovodGlobalState& state, MPIContext& ctx, bool is_coordinator);
+
+#if HAVE_MLSL
+static void server_affinity_set(int affinity) {
+  cpu_set_t cpuset;
+  pthread_t current_thread = pthread_self();
+
+  __CPU_ZERO_S(sizeof(cpu_set_t), &cpuset);
+  __CPU_SET_S(affinity, sizeof(cpu_set_t), &cpuset);
+
+  if (pthread_setaffinity_np(current_thread, sizeof(cpu_set_t), &cpuset) != 0) {
+    LOG(ERROR) << "setaffinity failed";
+  }
+
+  // Check if we set the affinity correctly
+  if (pthread_getaffinity_np(current_thread, sizeof(cpu_set_t), &cpuset) != 0) {
+    LOG(ERROR) << "sched_getaffinity failed";
+  }
+
+  for (int core_idx = 0; core_idx < __CPU_SETSIZE; core_idx++) {
+    if (__CPU_ISSET_S(core_idx, sizeof(cpu_set_t), &cpuset)) {
+      LOG(DEBUG) << "BG-thread affinity " << core_idx;
+    }
+  }
+}
+#endif
+
 void BackgroundThreadLoop(HorovodGlobalState& state, MPIContext& ctx) {
+#if HAVE_MLSL
```

Comment:
I feel that this adds a lot of complication to the `BackgroundThreadLoop`.  I'm wondering if some of this can be pulled out into `MLSLContext` in some way?  Background thread initialization is something that will ultimately need to be abstracted between MLSL, MPI, Gloo, etc.

Reply:
Moved init and finalize logic to MLSLContext